sympy/core/operations.py

```diff
@@ -611,23 +611,12 @@ def dispatch(self, types):
             if i == 0:
                 result = tup
             else:
-                selected = self.binary_select(result[0], tup[0])
+                selected = self.dispatcher.dispatch(result[0], tup[0])(result[0], tup[0])
                 if selected is tup[0]:
                     result = tup
                 elif selected is result[0]:
                     pass
```

Comment:
Hmm, this logic looks wrong to me.
I think it's fairly important to support something like:
```python
class Foo(Expr):
    def _add_handler(...): ...

class Bar(Expr):
    def _add_handler(...): ...

class FooBar(Foo, Bar):
   """ knows about both foo and bar, supports the methods of both at the same time """
   def _add_handler(...):  ...

@add.register_priority(Foo, Bar)
def _(a, b):
    return FooBar
```

Reply:
`_(a, b)` must return instance (not `FooBar` class) for recursive comparison because dispatcher checks type. With your design, `add(Foo(1), Bar(2), Spam(3))` will:
1. Compare type of `Foo(1)` (=`Foo` class) and type of `Bar(2)` (=`Bar` class), select `FooBar` class
2. Compare type of `FooBar` class (=`ManagedProperties`) and type of `Spam(3)` (=`Spam`) ...?

Reply:
Plus, allowing third option where you are doing binary selection seems treacherous, especially when arguments are compared recursively. Do we have to register `(Foo, Egg)` for `add(Foo(1), Egg(2))`, `(Bar, Egg)` for `add(Bar(1), Egg(2))` and yet again register `(FooBar, Egg)` for `add(Foo(1), Bar(2), Egg(3))`? It will get even perplexing if  more and more classes - `Bacon`, `Sausage`, and `Spam`- follow this design.

Reply:
> `_(a, b)` must return instance (not `FooBar` class)

That's not what your tests do. They both take and return the class itself. I think this part of your statement is wrong.

Reply:
Oh, geez. I'm sorry. You're right. I confused with older design.

Reply:
> Do we have to register `(Foo, Egg)` for `add(Foo(1), Egg(2))`, `(Bar, Egg)` for `add(Bar(1), Egg(2))` and yet again register `(FooBar, Egg)` for `add(Foo(1), Bar(2), Egg(3))`?

If you don't register `(FooBar, Egg)`, both `(Bar, Egg)` and `(Foo, Egg)` will match, and you'll get a (silenced) ambiguity warning.

> It will get even perplexing if more and more classes - Bacon, Sausage, and Spam- follow this design.

Right, but that complexity falls on the classes using the dispatch. The whole idea behind this type of multiple dispatch is to  allow different types to interact with each other.
Also, I have a concrete use case for this. I have:
* `DifferentialOperator`, a type that overloads `mul` to mean "take the derivative and do some stuff"
* `MultiVector`, a type that that implements the `|` operator (and overloads `mul` to return more `MultiVector`s)
* `MultiVectorDifferentialOperator`, which does both of the above at once.

For that to work for me, I need to be able to register `(DiffOp, MV)` to return `MvDiffOp`.

I don't see any reason not to allow that - it doesn't really add complexity to your dispatcher.